### ISSUE TITLE
Bug: [no-useless-assignment] false positive: assumes that let variable is not used

### ISSUE THREAD

**Issue by stanimirgeorgiev** [eslint/eslint#20846]
### Environment

ESLint Playground

### What parser are you using?

@typescript-eslint/parser

### Playground Code
```js
/* eslint no-useless-assignment: "error" */
const createCode = (number) => {
    let code = '';

    if (number > 1) {
        code = `has: ${number}`;
    } else {
        return undefined;
    }

    return {
        code: code,
    };
};

main();

```

### ESLint Configuration 
<details> 
 <summary> Tap to see full config </summary> 

```js
import { defineConfig } from "eslint/config";

export default defineConfig([
    {}
]);
```
 </details>

### ESLint Output
```
3:9 The value assigned to 'code' is not used in subsequent statements. (no-useless-assignment)
```

### What did you expect to happen?

No errors from ESLint.

### What actually happened?

3:9 The value assigned to 'code' is not used in subsequent statements. (no-useless-assignment)

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVzZWxlc3MtYXNzaWdubWVudDogXCJlcnJvclwiICovXG5jb25zdCBjcmVhdGVDb2RlID0gKG51bWJlcikgPT4ge1xuICAgIGxldCBjb2RlID0gJyc7XG5cbiAgICBpZiAobnVtYmVyID4gMSkge1xuICAgICAgICBjb2RlID0gYGhhczogJHtudW1iZXJ9YDtcbiAgICB9IGVsc2Uge1xuICAgICAgICByZXR1cm4gdW5kZWZpbmVkO1xuICAgIH1cblxuICAgIHJldHVybiB7XG4gICAgICAgIGNvZGU6IGNvZGUsXG4gICAgfTtcbn07XG5cbm1haW4oKTtcbiIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==

### Participation

- [ ] I am willing to submit a pull request for this issue.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

If I don't initialize the variable `code` there is no error

**Comment by mdjermanovic** [comment]
The `no-useless-assignment` rule works as intended in this case because variable `code` is either reassigned in the `if` branch (`` code = `has: ${number}` ``) or otherwise not read since the `else` branch exits the function. Overall, the empty string value assigned to it on initialization is not used in any code paths and thus the initialization can be removed. So, the fix would be to change `let code = '';` to `let code;`
